crates/sparrow-backend/src/exprs/expr_pattern.rs

```diff
@@ -0,0 +1,166 @@
+use std::str::FromStr;
+
+use smallvec::{smallvec, SmallVec};
+use sparrow_arrow::scalar_value::ScalarValue;
+
+use crate::exprs::expr_lang::ExprLang;
+use crate::exprs::ExprVec;
+use crate::Error;
+
+/// A representation of an expression with "holes" that may be instantiated.
+#[derive(Debug, Default)]
+pub(crate) struct ExprPattern {
+    pub(super) expr: egg::PatternAst<ExprLang>,
+}
+
+#[static_init::dynamic]
+pub(crate) static INPUT_VAR: egg::Var = egg::Var::from_str("?input").unwrap();
+
+impl ExprPattern {
+    /// Create a new `ExprPattern` which is an identity expression referencing the input.
+    pub fn new_input() -> error_stack::Result<Self, Error> {
+        let mut exprs = ExprPattern::default();
+        exprs.add_var(*INPUT_VAR)?;
+        Ok(exprs)
+    }
+
+    /// Create an expr pattern containing the given arguments.
+    ///
+    /// Returns the resulting pattern as well as the IDs of each of the arguments.
+    pub fn new_instruction(
+        name: &'static str,
+        literal_args: smallvec::SmallVec<[ScalarValue; 2]>,
+        args: Vec<ExprPattern>,
+        data_type: arrow_schema::DataType,
+    ) -> error_stack::Result<ExprPattern, Error> {
+        let mut graph = egg::EGraph::<egg::ENodeOrVar<ExprLang>, ()>::default();
+        let mut arg_ids = SmallVec::with_capacity(args.len());
+        for arg in args {
+            let id = graph.add_expr(&arg.expr);
+            arg_ids.push(id);
+        }
+
+        // We can only extract a single expression, so we create one.
+        // This is why we need to know the instruction to create, rather than
+        // just returning the resulting `egg::Id` for each argument.
+        let output = graph.add(egg::ENodeOrVar::ENode(ExprLang {
+            name,
+            literal_args,
+            args: arg_ids,
+            result_type: data_type,
+        }));
+
+        let cost_function = egg::AstSize;
+        let extractor = egg::Extractor::new(&graph, cost_function);
+        let (_best_cost, expr) = extractor.find_best(output);
+
+        Ok(ExprPattern { expr })
+    }
+
+    /// Instantiate the pattern.
+    ///
+    /// Replaces `?input` with the `input` instruction.
+    pub fn instantiate(
+        &self,
+        input_type: arrow_schema::DataType,
+    ) -> error_stack::Result<ExprVec, Error> {
+        let mut graph = egg::EGraph::<ExprLang, ()>::default();
+
+        let input_id = graph.add(ExprLang {
+            name: "input",
+            literal_args: smallvec![],
+            args: smallvec![],
+            result_type: input_type,
+        });
+        let mut subst = egg::Subst::with_capacity(1);
+        subst.insert(*INPUT_VAR, input_id);
+
+        let result = graph.add_instantiation(&self.expr, &subst);
+
+        let cost_function = egg::AstSize;
+        let extractor = egg::Extractor::new(&graph, cost_function);
+        let (_best_cost, expr) = extractor.find_best(result);
```

Comment:
Is there a reason `find_best` is needed here? Is it not just a single substitution in the graph here?

Reply:
We could instantiate the pattern `ExprPattern` to make an `ExprVec`, but it would require iterating over all of teh nodes in the pattern manually. Using the graph let's us use `add_instantiation` (from EGraph) and also provides de-duplication, etc.